crates/storage/provider/src/transaction/mod.rs

```diff
@@ -0,0 +1,740 @@
+        let hashed = storages.into_iter().fold(BTreeMap::new(), |mut map, (address, storage)| {
+            let storage = storage.into_iter().fold(BTreeMap::new(), |mut map, (key, value)| {
+                map.insert(keccak256(key), value);
+                map
+            });
+            map.insert(keccak256(address), storage);
+            map
+        });
+
+        let mut hashed_storage = self.cursor_dup_write::<tables::HashedStorage>()?;
+        // Hash the address and key and apply them to HashedStorage (if Storage is None
+        // just remove it);
+        hashed.into_iter().try_for_each(|(address, storage)| {
+            storage.into_iter().try_for_each(|(key, value)| -> Result<(), TransactionError> {
+                if hashed_storage
+                    .seek_by_key_subkey(address, key)?
+                    .filter(|entry| entry.key == key)
+                    .is_some()
+                {
+                    hashed_storage.delete_current()?;
+                }
+
+                if value != U256::ZERO {
+                    hashed_storage.upsert(address, StorageEntry { key, value })?;
+                }
+                Ok(())
+            })
+        })?;
+        Ok(())
+    }
+
+    /// Iterate over account changesets and return all account address that were changed.
+    pub fn get_addresses_of_changed_accounts(
+        &self,
+        from: TransitionId,
+        to: TransitionId,
+    ) -> Result<BTreeSet<Address>, TransactionError> {
+        Ok(self
+            .cursor_read::<tables::AccountChangeSet>()?
+            .walk_range(from..to)?
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            // fold all account to one set of changed accounts
+            .fold(BTreeSet::new(), |mut accounts: BTreeSet<Address>, (_, account_before)| {
+                accounts.insert(account_before.address);
+                accounts
+            }))
+    }
+
+    /// Get plainstate account from iterator
+    pub fn get_plainstate_accounts(
+        &self,
+        iter: impl IntoIterator<Item = Address>,
+    ) -> Result<Vec<(Address, Option<Account>)>, TransactionError> {
+        let mut plain_accounts = self.cursor_read::<tables::PlainAccountState>()?;
+        Ok(iter
+            .into_iter()
+            .map(|address| plain_accounts.seek_exact(address).map(|a| (address, a.map(|(_, v)| v))))
+            .collect::<Result<Vec<_>, _>>()?)
+    }
+
+    /// iterate over accounts and insert them to hashing table
+    pub fn insert_account_for_hashing(
+        &self,
+        accounts: impl IntoIterator<Item = (Address, Option<Account>)>,
+    ) -> Result<(), TransactionError> {
+        let mut hashed_accounts = self.cursor_write::<tables::HashedAccount>()?;
+
+        let hashes_accounts = accounts.into_iter().fold(
+            BTreeMap::new(),
+            |mut map: BTreeMap<H256, Option<Account>>, (address, account)| {
+                map.insert(keccak256(address), account);
```

Comment:
This feels leaky to me, I think a better abstraction is having types that can be written to the database instead (like the `account.apply_to_db` further down). Or at least this part with hashing etc should not be in the transaction.

Another thing is that moving everything here gives us more overhead because we are constantly opening new cursors

Reply:
Yeah, I feel the same, we are putting a lot of things in one structure and just growing it bigger and bigger, it is not nice.

I still don't have a killer pattern here that would encompass all behaviour that we need. 
What would be great is to have abstraction that can work on `Database`, `Transaction` and `Cursor` and their references. That could be used from both `SharedDatabase` and `Transaction` ("ContinuousTransaction") as we have a need for both of them.

I experimented with a few of them.
`StateProviders` for example takes a generic transaction and does stuff on it, the problem there is that reference implementation has boilerplate code, maybe we can implement DbTx trait on `&DbTx`  that would make this a lot insier. I did something similar to `Database` trait and relaxed some constraints. This would take data in function and does the whatever is needed.
The second example is what you mentioned with `account.apply_to_db` is to add function that takes generic Cursor/Tx and applies to code to it.  This is the inverse of first example where data is already there and we are sending transaction inside function. Downside is you would need to create object for every functionality you need. 
One time I wanted to tie the functionality to the table or multiple tables, but I didn't find a way how to do it.

Either way, yeah.

Reply:
Is ETL going to make this a lot simpler?

Reply:
It is not about `ETL` but about code patterns and how we are inserting or getting data. ETL is another patter that we need to think about. `StateProvider` for example doesn't have anything with `ETL` but it is a helper for accessing database.